packages/fuels-contract/src/script.rs

```diff
@@ -178,86 +214,107 @@ impl Script {
-                ids
-            })
-            .collect();
-
-        // We must associate the right external contract input to the corresponding external
-        // output index (TXO)
-        for (idx, contract_id) in contract_ids.into_iter().enumerate() {
-            let zeroes = Bytes32::zeroed();
-            let self_contract_input = Input::contract(
-                UtxoId::new(Bytes32::zeroed(), idx as u8),
-                zeroes,
-                zeroes,
-                contract_id,
-            );
-            inputs.push(self_contract_input);
-
-            let external_contract_output = Output::contract(idx as u8, zeroes, zeroes);
-            outputs.push(external_contract_output);
-        }
+        (inputs, outputs)
+    }
 
-        // Get all unique asset ids
-        let asset_ids: HashSet<AssetId> = calls
+    fn extract_unique_asset_ids(spendable_coins: &[Coin]) -> HashSet<AssetId> {
+        spendable_coins
             .iter()
-            .map(|call| call.call_parameters.asset_id)
-            .chain(iter::once(AssetId::default()))
-            .collect();
+            .map(|coin| coin.asset_id.clone().into())
+            .collect()
+    }
 
-        let mut spendables = vec![];
-        for asset_id in asset_ids.iter() {
-            spendables.extend(
-                wallet
-                    .get_spendable_coins(*asset_id, DEFAULT_SPENDABLE_COIN_AMOUNT as u64)
-                    .await
-                    .unwrap(),
-            );
-        }
+    fn extract_variable_outputs(calls: &[ContractCall]) -> Vec<Output> {
+        calls
+            .iter()
+            .filter_map(|call| call.variable_outputs.clone())
+            .flatten()
+            .collect()
+    }
 
-        for asset_id in asset_ids.iter() {
-            // add asset change if any inputs are being spent
-            let change_output = Output::change(wallet.address().into(), 0, asset_id.to_owned());
-            outputs.push(change_output);
-        }
+    fn generate_asset_change_outputs(
+        wallet_address: &Bech32Address,
+        asset_ids: HashSet<AssetId>,
+    ) -> Vec<Output> {
+        asset_ids
+            .into_iter()
+            .map(|asset_id| Output::change(wallet_address.into(), 0, asset_id))
+            .collect()
+    }
 
-        for coin in spendables {
-            let input_coin = Input::coin_signed(
-                UtxoId::from(coin.utxo_id),
-                coin.owner.into(),
-                coin.amount.0,
-                coin.asset_id.into(),
-                0,
-                0,
-            );
+    fn generate_contract_outputs(num_of_contracts: usize) -> Vec<Output> {
+        (0..num_of_contracts)
+            .map(|idx| Output::contract(idx as u8, Bytes32::zeroed(), Bytes32::zeroed()))
+            .collect()
+    }
 
-            inputs.push(input_coin);
-        }
+    fn convert_to_signed_coins(spendable_coins: Vec<Coin>) -> Vec<Input> {
```

Comment:
One of the variants of Input is Input::CoinSigned, not sure what that is but seemed to me like it could be rephrased as a 'signed coin' or 'signed coin input'.